engine/deeplearning/src/main/java/org/opennms/oce/engine/deeplearning/TFClusterer.java

```diff
@@ -0,0 +1,417 @@
+    private final Vectorizer vectorizer;
+
+    private final double epsilon;
+    private final int numGraphThreads;
+    private final int numTfThreads;
+
+    private ExecutorService graphExecutor;
+    private ExecutorService tfExecutor;
+
+    private final WeakComponentClusterer<CEVertex, CEEdge> weakComponentClusterer = new WeakComponentClusterer<>();
+
+    public TFClusterer(TFModel tfModel, Vectorizer vectorizer, DeepLearningEngineConf conf) {
+        this.tfModel = Objects.requireNonNull(tfModel);
+        this.vectorizer = Objects.requireNonNull(vectorizer);
+        Objects.requireNonNull(conf);
+
+        epsilon = conf.getEpsilon();
+        numGraphThreads = conf.getNumGraphProcessingThreads();
+        numTfThreads = conf.getNumTensorFlowProcessingThreads();
+    }
+
+    public void init() {
+        graphExecutor = Executors.newFixedThreadPool(numGraphThreads, new ThreadFactoryBuilder()
+                .setNameFormat("tf-clusterer-graph-%d")
+                .build());
+        tfExecutor = Executors.newFixedThreadPool(numTfThreads, new ThreadFactoryBuilder()
+                .setNameFormat("tf-clusterer-tf-%d")
+                .build());
+    }
+
+    public void destroy() {
+        graphExecutor.shutdown();
+        tfExecutor.shutdown();
+    }
+
+    /**
+     * Cluster the alarms on the given graph:
+     *
+     * 1. Gather the vertices with alarms
+     * 2. Split the graph into disconnected subgraphs and process these in parallel
+     * 3. Skip graphs without any alarms
+     * 4. For every vertex with alarms:
+     * 4.a) Compare all of the alarms on that vertex
+     * 4.b) Find and compare alarms on all other vertices within an epsilon radius
+     *
+     * We use two different thread pools to accomplish this.
+     *
+     * Threads in the graph processing pool are used to traverse the graph and
+     * match candidate vertices. When alarms on these vertices need to be matched
+     * a task in placed on a queue.
+     *
+     * Threads in the TensorFlow processing pool consume and process the "pairing"
+     * tasks generated the by graph processing. When pairs are matched successfully,
+     * the result in placed on a different queue.
+     *
+     * The main thread (caller) processes the matches to build clusters incrementally
+     * as the results are available and will remain blocked until all of the tasks have
+     * been completed.
+     *
+     * Further optimizations include:
+     *  * Work to avoid processing alarms that are already in clusters
+     *  * Cache previous results?
+     *  * Find additional ways of limiting the number of comparisions
+     *
+     * @param g graph with alarms to cluster
+     * @return clusters of alarms
+     */
+    public List<Cluster<AlarmInSpaceTime>> cluster(Graph<CEVertex, CEEdge> g) {
+        // Gather the list of vertices with alarms
+        final Set<CEVertex> verticesWithAlarms = new LinkedHashSet<>();
+        for (CEVertex v : g.getVertices()) {
```

Comment:
I generally try and avoid using `.parallelStream()` since I'm not sure what else is using the threads in the common `ForkJoinPool`.

I'm thinking that we may be able to get rid of this block altogether and delegate it to the cluster engine instead - i.e. track these incrementally instead of searching every time.

Will leave as it for now.

Reply:
Yeah, you would need to provide it with an explicit thread pool. However I'm thinking the performance gain is probably extremely minimal unless the number of vertices is huge since `v.hasAlarms()` is probably trivially quick.